Generalize performance queries

arman-ddl opened the pull request:
Will un-draft after Questions are processed.

### Notes
N1) Created extensions-performance module and moved `PerformanceQueries` code there to prevent circular dependency
N2) The preexisting `PerformanceQueries.queryUpdatePerformanceMap` was broken b/c of `groupBy`'ing a formatted table; I've made formatting optional in `queryUpdatePerformance` and in general deferred formatting until table processing is complete

### Questions
Q1) What does "EvaluationNumber" mean? Is it the same across QPL, QOPL, and UPL? The docs seem to suggest a different meaning than the code: "... the EvaluationNumber column tells how many times each process has been evaluated." I ask b/c I want to make sure all processing is still meaningful in cases that are not filtering by evaluation number; in particular, in `queryUpdatePerformanceMap`.
Q2) Is `Classpaths.inheritGroovy(project, 'groovy', 'implementation')` needed in the new module? Are there other Gradle changes I should make?

devinrsmith reviewed (changes requested):
> N1) Created extensions-performance module and moved PerformanceQueries code there to prevent circular dependency

It looks like we still have a circular-ish dependency:

engine-table -> extensions-performance (runtimeOnly)
extensions-performance -> engine-table (implementation)

Even though gradle / java doesn't technically complain, it's still not a good design.

What we probably want to do is to move the `runtimeOnly project(':extensions-performance')` to `server/jetty-app/build.gradle`.

> N2) The preexisting PerformanceQueries.queryUpdatePerformanceMap was broken b/c of groupBy'ing a formatted table; I've made formatting optional in queryUpdatePerformance and in general deferred formatting until table processing is complete

We should probably have tests if we want to ensure it continues to work in the future. cc @rcaudy

> Q1) What does "EvaluationNumber" mean?

cc @rcaudy 

> Q2) Is Classpaths.inheritGroovy(project, 'groovy', 'implementation') needed in the new module? Are there other Gradle changes I should make?

No, we don't need it; really, we are just using the groovy file as a resource, not depending on groovy at compile time in any way. Remove it.

We should also probably just use `src/main/resources` instead of hacking in `groovy` as a resource source set.

We should also move the InitScript for "1-performance.groovy" into this new module. Although, I'm confused b/c I see two of them now that I am looking:

`io.deephaven.engine.util.GroovyDeephavenSession.PerformanceQueries`
`io.deephaven.engine.util.PerformanceScript`

arman-ddl commented:
* Moved runtime dep to jetty-app
* I moved the groovy into resources and removed `sourceSets` logic; fwiw, this style is also used in engine-table (which I was mimicking when I split code off from there)
* I moved GroovyDeephavenSession.PerformanceQueries to PerformanceQueries.InitScript


Also, I tried removing `inheritGroovy` from extensions-performance, but that actually fails `prepareCompose` w/:
```
/.../deephaven-core/extensions/performance/src/main/java/io/deephaven/engine/table/impl/util/PerformanceQueriesGeneral.java:279: error: cannot access Closure
                .plot("UGPRatio", pm, "IntervalEnd", "UGPTimeRatio")
                ^
  class file for groovy.lang.Closure not found
```

devinrsmith commented on extensions/performance/src/main/java/io/deephaven/engine/table/impl/util/PerformanceQueries.java:
```diff
@@ -0,0 +1,173 @@
+        return PerformanceQueriesGeneral.serverStateWithPlots(TableLoggers.serverStateLog());
+    }
+
+    @AutoService(GroovyDeephavenSession.InitScript.class)
+    public static class InitScript implements GroovyDeephavenSession.InitScript {
```
I think we can delete `Plot/src/main/java/io/deephaven/engine/util/PerformanceScript.java`, I'm guessing it was accidentally copied originally.

devinrsmith commented on server/jetty-app/build.gradle:
```diff
@@ -10,6 +10,7 @@ configurations {
 dependencies {
     implementation project(':server-jetty')
 
+    runtimeOnly project(':extensions-performance')
```
Given that server already has a dependency on this, we don't need to specify this dependency here.

devinrsmith commented on engine/table/build.gradle:
```diff
@@ -90,7 +90,8 @@ dependencies {
     Classpaths.inheritAssertJ(project)
 
-    testRuntimeOnly project(':log-to-slf4j'),
+    testRuntimeOnly project(':extensions-performance'),
```
What tests fail when this dependency is missing?

arman-ddl replied:
All TestGroovyDeephavenSession tests. Example of stack trace:
```
io.deephaven.engine.util.GroovyExceptionWrapper: startup failed:
Script_1: 4: unable to resolve class io.deephaven.engine.table.impl.util.PerformanceQueries
 @ line 4, column 1.
   import static io.deephaven.engine.table.impl.util.PerformanceQueries.*;
   ^

1 error

	at app//org.codehaus.groovy.control.ErrorCollector.failIfErrors(ErrorCollector.java:292)
	at app//org.codehaus.groovy.control.CompilationUnit$ISourceUnitOperation.doPhaseOperation(CompilationUnit.java:914)
	at app//org.codehaus.groovy.control.CompilationUnit.compile(CompilationUnit.java:627)
	at app//io.deephaven.engine.util.GroovyDeephavenSession.updateClassloader(GroovyDeephavenSession.java:556)
	at app//io.deephaven.engine.util.GroovyDeephavenSession.evaluate(GroovyDeephavenSession.java:225)
	at app//io.deephaven.engine.util.AbstractScriptSession.lambda$evaluateScript$1(AbstractScriptSession.java:145)
	at app//io.deephaven.engine.context.ExecutionContext.lambda$apply$0(ExecutionContext.java:129)
	at app//io.deephaven.engine.context.ExecutionContext.apply(ExecutionContext.java:140)
	at app//io.deephaven.engine.context.ExecutionContext.apply(ExecutionContext.java:128)
	at app//io.deephaven.engine.util.AbstractScriptSession.evaluateScript(AbstractScriptSession.java:145)
	at app//io.deephaven.engine.util.ScriptSession.evaluateScript(ScriptSession.java:113)
	at app//io.deephaven.engine.util.scripts.TestGroovyDeephavenSession.testNullCast(TestGroovyDeephavenSession.java:57)
	at java.base@11.0.17/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base@11.0.17/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base@11.0.17/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base@11.0.17/java.lang.reflect.Method.invoke(Method.java:566)
	at app//org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at app//org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at app//org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at app//org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at app//org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at app//org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at app//org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at app//org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at app//org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at app//org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at app//org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at app//org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at app//org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at app//org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at app//org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at app//org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at app//org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at app//org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.runTestClass(JUnitTestClassExecutor.java:108)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:57)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:39)
	at org.gradle.api.internal.tasks.testing.junit.AbstractJUnitTestClassProcessor.processTestClass(AbstractJUnitTestClassProcessor.java:62)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.processTestClass(SuiteTestClassProcessor.java:52)
	at java.base@11.0.17/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base@11.0.17/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base@11.0.17/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base@11.0.17/java.lang.reflect.Method.invoke(Method.java:566)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
	at com.sun.proxy.$Proxy2.processTestClass(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker$2.run(TestWorker.java:176)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.executeAndMaintainThreadName(TestWorker.java:129)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:100)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:60)
	at org.gradle.process.internal.worker.child.ActionExecutionWorker.execute(ActionExecutionWorker.java:56)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:113)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:65)
	at app//worker.org.gradle.process.internal.worker.GradleWorkerMain.run(GradleWorkerMain.java:69)
	at app//worker.org.gradle.process.internal.worker.GradleWorkerMain.main(GradleWorkerMain.java:74)
```

devinrsmith replied:
Looks like this is an implicit dependency that is brought in from `io.deephaven.engine.util.GroovyDeephavenSession#fullCommand`. I'm hoping we can remove it without ill-effect?

devinrsmith commented on extensions/performance/src/main/java/io/deephaven/engine/table/impl/util/PerformanceQueriesGeneral.java:
```diff
@@ -1,53 +1,37 @@
+ * Generalizes {@link PerformanceQueries} to accept table parameters and make evaluation number parameter optional.
+ */
+public class PerformanceQueriesGeneral {
+    private static boolean formatPctColumns = true;
```
Is your intention to set this to `false` in DHE, or in specific cases? It's a bit fragile. Have you considered adding a boolean parameter to all of the methods?

It looks like a few of the methods have formatPctColumnsLocal; what is the global static for?

arman-ddl replied:
I decided not to set `formatPctColumns` from true to false to true as needed. Fwiw, this static field already existed despite never being set to anything other than initial value.

Instead, I added `formatPctColumnsLocal` where it was _needed_ (meaning to avoid the bug referenced in N2 via good practices reg order of operations). In this case, both the local and static value need to be true to allow formatting. I don't know the purpose of static `formatPctColumns`, but did not want to remove it w/o a DHC dev's say-so.
